adage/analyze/api.py

```diff
@@ -427,7 +427,14 @@ def apply_filters(self, request, applicable_filters):
                 raise BadRequest("Invalid gene IDs: %s" % genes)
             # "in" operator in Django supports both list and set.
             qset = Q(gene1__in=ids) | Q(gene2__in=ids)
-            object_list = object_list.filter(qset).distinct()
+            direct_edges = Edge.objects.filter(qset).distinct()
+            related_genes = set()
+            for e in direct_edges:
+                related_genes.add(e.gene1)
+                related_genes.add(e.gene2)
+            object_list = object_list.filter(
+                gene1__in=related_genes, gene2__in=related_genes
+            ).distinct()
```

Comment:
Let me see if I follow: before, we applied `qset` as a filter to remove `Edge` objects that did not have either `gene1` or `gene2`. The issue is raised because this is too restrictive: we are *also* filtering out edges that are present in the network between genes that share an edge with our search set in `genes` because those are not direct edges.

So, now, we start by creating a list of `direct_edges`, expand out to get all of the genes that participate in those edges (`related_genes`) and *then* we go back and filter our overall `Edge` list to include only those edges touching one of those `related_genes`.

This seems like a logical approach, but I don't think it matches what was requested. Why, for example, are we seeing many more *genes* than the original query, when we were just looking for more *edges*?

I think I can answer. The new `object_list.filter` applied at lines 435-437 is picking out all edges that touch one of the `related_genes`. Since we have already expanded our gene list to get `related_genes`, what we actually want here is to pick out *only* edges where both genes are in our list of `related_genes`. I think you could express this via a new `qset` similar to line 429:

```python
qset = Q(gene1__in=related_genes) & Q(gene2__in=related_genes)
```

and then we could use this new `qset` in place of the filter at lines 435-437:

```python
object_list = object_list.filter(qset).distinct()
```
See what I am saying? I'm taking a small leap in the syntax by assuming that the `&` operator will do what I expect because I haven't fully read up on `django.db.models.Q`, but do you follow my logic?